### PR TITLE
Make scratch build names unique

### PR DESCRIPTION
**Pull request by csomh**
As of now two scratch builds could have the same name,
due to the fact that build names were generated using a timestamp.

This changes is adding a random 5 digit number to the build names,
which should ensure uniqueness of the build names.

### CODE REVIEW

File: osbs/api.py [hunk 1]
```diff
@@ -305,7 +306,9 @@ def _create_scratch_build(self, build_request):
             builder_img['kind'] = 'DockerImage'
             builder_img['name'] = ref
 
-        build_config_name = 'scratch-%s' % datetime.datetime.now().strftime('%Y%m%d%H%M%S')
+        build_config_name = 'scratch-%s-%s' % (
```
Comment (vrutkovs) [inline osbs/api.py]: Maybe ```import uuid``` and ```str(uuid.uuid1())[:8]``` would be simpler?

**Comment by lcarva** [comment]
Actually, I'm not sure anymore if this will not also update the unique tag used. Could you verify it?

**Comment by twaugh** [comment]
Need to update this part I think: https://github.com/projectatomic/osbs-client/blob/master/osbs/build/spec.py#L300

File: osbs/api.py (hunk 1 above)
Comment (csomh) [inline osbs/api.py]: `uuid.uuid1()` seems to be doing much more then `random`:

```
Generate a UUID from a host ID, sequence number, and the current time.
If 'node' is not given, getnode() is used to obtain the hardware
address.  If 'clock_seq' is given, it is used as the sequence number;
otherwise a random 14-bit sequence number is chosen.
```
Would there be any other benefits of using it?

Also: switched to using `randrange` instead of `randint`, which had the strange behavior to include the end point.

**Comment by csomh** [comment]
Unit tests added. Also tried to solve #483 (unique tags), but somewhat got lost between BuildRequest and BuildSpec usage and how this design should work.

Anyways, this should ensure tag uniqueness and reuses the same unique string to update build names for scratch builds, by reusing the last 20 chars from the output name.

Let me know what you think about it.

File: osbs/api.py [hunk 2]
```diff
@@ -305,7 +302,8 @@ def _create_scratch_build(self, build_request):
             builder_img['kind'] = 'DockerImage'
             builder_img['name'] = ref
 
-        build_config_name = 'scratch-%s' % datetime.datetime.now().strftime('%Y%m%d%H%M%S')
+        build_config_name = 'scratch-%s' % (
+            build_json['spec']['output']['to']['name'][-20:])
```
Comment (vrutkovs) [inline osbs/api.py]: 20 here is a magic number - lets not use those or at least use a constant for it.

Why is it even set to 20? Is this the registry hostname part?

File: tests/build_/test_build_request.py [hunk 3]
```diff
@@ -205,7 +205,7 @@ def test_render_simple_request(self, registry_uris, build_image):
         assert build_json["spec"]["source"]["git"]["uri"] == TEST_GIT_URI
         assert build_json["spec"]["source"]["git"]["ref"] == TEST_GIT_REF
 
-        expected_output = "john-foo/component:none-20"
+        expected_output = "john-foo/component:none-"
```
Comment (vrutkovs) [inline tests/build_/test_build_request.py]: Tests should mock `random.randrange` and check that the full image is verified

File: osbs/api.py [hunk 4]
```diff
@@ -289,8 +288,6 @@ def _create_scratch_build(self, build_request):
         logger.debug(build_request)
         build_json = build_request.render()
         build_json['kind'] = 'Build'
-        if 'spec' not in build_json.keys():
```
Comment (lcarva) [inline osbs/api.py]: Why was this removed? Seems unrelated.

File: osbs/api.py (hunk 2 above)
Comment (lcarva) [inline osbs/api.py]: `spec.output.to.name` refers to the image that this build should produce. This seems like an attempt to re-use the timestamp and random string used there. I like that, but it's not very clear from the code and it seems somewhat brittle.

How about something like this:
```python
output_image_name = build_json['spec']['output']['to']['name']
# Reuse random string and timestamp values.
build_config_name = 'scratch-%s-%s' % tuple(output_image_name.rsplit('-', 2)[-2:])
```

Comment (csomh) [inline osbs/api.py]: Yes, this was the intent. Updated like recommended to make it easier to understand. Thanks!

File: tests/build_/test_build_request.py (hunk 3 above)
Comment (csomh) [inline tests/build_/test_build_request.py]: `random.randrange` is called in `BuildSpec`, so it's tested there. Updated those tests to mock both this and `utcnow()`.

For this test, I think it's enough to check only `BuildRequest` functionality (that output name is prefixed with `registry_uris[0]`, when there is one).

File: osbs/api.py (hunk 4 above)
Comment (csomh) [inline osbs/api.py]: I just stumbled upon this, and yes it's unrelated. See comment in commit msg:

```
Remove check for 'spec' from _create_scratch_build

After rendering the build_request, a check was made
to see if the rendered build_json has 'spec'.

This is redundant, as build_request objects already have
'spec' in them.
```

**Comment by twaugh** [comment]
Does commit 61c83ad4a1f298b210e4e00b62bea6402dc75c66 need squashing into some other commit before this is merged?

**Comment by csomh** [comment]
Squashed and rebased.
